sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -584,24 +602,39 @@ class JoinSuite extends QueryTest with SharedSQLContext {
     val cartesianQueries = Seq(
       /** The following should error out since there is no explicit cross join */
       "SELECT * FROM testData inner join testData2",
-      "SELECT * FROM testData left outer join testData2",
-      "SELECT * FROM testData right outer join testData2",
-      "SELECT * FROM testData full outer join testData2",
+
       "SELECT * FROM testData, testData2",
       "SELECT * FROM testData, testData2 where testData.key = 1 and testData2.a = 22",
+      "SELECT * FROM testData, testData2 where testData.key > testData2.a",
+      "SELECT * FROM testData INNER JOIN testData2 ON testData.key + testData2.a = testData2.b",
       /** The following should fail because after reordering there are cartesian products */
       "select * from (A join B on (A.key = B.key)) join D on (A.key=D.a) join C",
       "select * from ((A join B on (A.key = B.key)) join C) join D on (A.key = D.a)",
       /** Cartesian product involving C, which is not involved in a CROSS join */
-      "select * from ((A join B on (A.key = B.key)) cross join D) join C on (A.key = D.a)");
+      "select * from ((A join B on (A.key = B.key)) cross join D) join C on (A.key = D.a)")
```

Comment:
The estimated size of `testData`, `testData2`, and `testData3` are much larger than the broadcast threshold. Thus, it does not consider the cases when the tables are small enough to broadcast. Thus, when the tables are small enough to broadcast, should we still block users? In that case, the join is not expensive.

It affects the whole detection logics rewriting. Thanks! @rxin @hvanhovell @cloud-fan @sameeragarwal